src/api/root/handler.js

```diff
@@ -0,0 +1,183 @@
+ This is the official list of the Mojaloop project contributors for this file.
+ Names of the original copyright holders (individuals or organizations)
+ should be listed with a '*' in the first column. People who have
+ contributed from an organization can be listed under the organization
+ that actually holds the copyright for their contributions (see the
+ Gates Foundation organization for an example). Those individuals should have
+ their names indented and be marked with a '-'. Email address can be added
+ optionally within square brackets <email>.
+ * Gates Foundation
+ - Name Surname <name.surname@gatesfoundation.com>
+
+ * Lewis Daly <lewis@vesselstech.com>
+ --------------
+ ******/
+
+const P = require('bluebird')
+
+const packageJson = require('../../../package.json')
+const Db = require('../../lib/db')
+const Util = require('../../handlers/lib/utility')
+const Kafka = require('../../handlers/lib/kafka/index')
+const Enum = require('../../lib/enum')
+const TransferEventType = Enum.transferEventType
+const TransferEventAction = Enum.transferEventAction
+
+const statusEnum = {
+  OK: 'OK',
+  DOWN: 'DOWN'
+}
+
+const serviceName = {
+  datastore: 'datastore',
+  broker: 'broker',
+  sidecar: 'sidecar',
+  cache: 'cache'
+}
+
+/**
+ * getHealth
+ *
+ * Gets the health status of the service.
+ *
+ *
+ * @example:
+ * returns the health of related sub-services:
+ * {
+ *   "status": "OK",
+ *   "uptime": 0,
+ *   "started": "2019-05-31T05:09:25.409Z",
+ *   "versionNumber": "5.2.3",
+ *   "services": [
+ *     {
+ *       "name": "datastore",
+ *       "status": "OK"
+ *     }
+ *   ]
+ * }
+ *
+ *
+ * @param {*} request - the request object
+ * @param {*} h       - the handler object
+ */
+const getHealth = async function (request, h) {
+  // Default values
+  let statusCode = 200
+  let status = statusEnum.OK
+  let isHealthy = true
+  let services = []
+  let subServices = {}
+
+  const uptime = process.uptime() // in seconds by default
+  const startTimeDate = new Date(Date.now() - uptime)
+  const startTime = startTimeDate.toISOString()
+  const versionNumber = packageJson.version
+
+  try {
+    services = [
+      await getSubServiceHealth(serviceName.datastore),
+      await getSubServiceHealth(serviceName.broker)
+    ]
+    isHealthy = evaluateServiceHealth(services)
+    subServices = {
+      services
+    }
+  } catch (err) {
+    isHealthy = false
+  }
+
+  if (!isHealthy) {
+    statusCode = 502
+    status = statusEnum.DOWN
+  }
+
+  const response = {
+    status,
+    uptime,
+    startTime,
+    versionNumber,
+    ...subServices
+  }
+  return h.response(response).code(statusCode)
+}
+
+/**
+ * getSubServiceHealth
+ *
+ * Gets the health status of a sub-service. Takes advantage of existing
+ * connections to the service (doesn't implement a ping or )
+ */
+const getSubServiceHealth = async function (serviceName) {
+  switch (serviceName) {
+    case 'datastore': {
+      const tables = await Db._listTables()
```

Comment:
We should also consider the following approaches for the DB health checks:

1. Check the migration_lock table `select is_locked from migration_lock;`. If `is_locked` === `0` we are all good. This is the same check we use on the init-containers.
2 `SELECT version()`. This purely checks connectivity. This is what was used in the init-containers before we implemented the readiness check on 1. above.

The nice thing about option 1 is that it guarantees that the DB is in a good state for processing. However, the 2nd option should be ok since we are checking option 1 on the init-container anyway, and on the health end-point we are more concerned about the connectivity to the db.

Let me know your thoughts here.

Reply:
Yeah we can really go either way on this.

Is there a chance that in the future we might want to run migrations while the API servers are still alive for some reason (for example minimizing downtime between deployments)? In that case *1* might make more sense so that kubernetes takes the node out of service. 

The same thing goes for relying on init containers, perhaps if the init containers were to fail to unlock the migration lock (or some other random bug), and we relied on *2*, then we could get in a tricky situation where we are trying to connect to the database when we shouldn't be.

Reply:
> Is there a chance that in the future we might want to run migrations while the API servers are still alive for some reason (for example minimizing downtime between deployments)? In that case _1_ might make more sense so that kubernetes takes the node out of service.

That is a valid point. That could occur on an upgrade through Helm and assuming that there were changes to the migration scripts....this would cause the migration to run. However, that may not be a bad thing as per your comment since you probably don't want to access the DB while the migration is running. Also, I wouldn't recommend upgrading with this approach if the system is still busy processing "live transactions". It would require some scheduled downtime.

> The same thing goes for relying on init containers, perhaps if the init containers were to fail to unlock the migration lock (or some other random bug), and we relied on _2_, then we could get in a tricky situation where we are trying to connect to the database when we shouldn't be.

You are 100% correct. But in theory the init-container is a gatekeeper, so the main container should never start up since the migration lock is never released. Thus Kubernetes will never have the opportunity to ever run any health checks against the main container.